### ISSUE TITLE
IRemoteRequestHandler.InvokeAsync() never returns when using AddMessagePipeNamedPipeInterprocess 

### ISSUE THREAD

**Issue by jcoliz** [Cysharp/MessagePipe#99]
Hi, I'm having a bit of trouble bringing up MessagePipe for IPC-RPC using named pipes. I've attempted to use the code described in the README, but to no avail.

I have create two projects, `Client` and `Server` using `dotnet new worker`. Code is below. Server seems to run successfully, but never executes `IAsyncRequestHandler.InvokeAsync` . Client never returns from the call to `InvokeAsync`. Using MessagePipe 1.7.4, and .NET 6.0.

My apologies if I'm missing something obvious!

# Client

## Program.cs

```c#
using Client;
using MessagePipe;

IHost host = Host.CreateDefaultBuilder(args)
    .ConfigureServices(services =>
    {
        services.AddMessagePipe();
        services.AddMessagePipeNamedPipeInterprocess("sample-pipe", config => config.InstanceLifetime = InstanceLifetime.Singleton );
        services.AddHostedService<Worker>();
    })
    .Build();

host.Run();
```

## Worker.cs

```c#
using MessagePipe;

namespace Client;

public class Worker : BackgroundService
{
    private readonly ILogger<Worker> _logger;
    private readonly IRemoteRequestHandler<int, string> _remoteHandler;

    public Worker(ILogger<Worker> logger, IRemoteRequestHandler<int, string> remoteHandler)
    {
        _logger = logger;
        _remoteHandler = remoteHandler;
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        while (!stoppingToken.IsCancellationRequested)
        {
            var result = await _remoteHandler.InvokeAsync(1234);
            _logger.LogInformation("Received: {result}", result);
            await Task.Delay(1000, stoppingToken);
        }
    }
}
```

# Server

## Program.cs

```c#
using MessagePipe;
using Server;

IHost host = Host.CreateDefaultBuilder(args)
    .ConfigureServices(services =>
    {
        services.AddHostedService<Worker>();
        services.AddMessagePipe();
        services.AddMessagePipeNamedPipeInterprocess("sample-pipe");
    })
    .Build();

host.Run();

public class MyAsyncHandler : IAsyncRequestHandler<int, string>
{
    public async ValueTask<string> InvokeAsync(int request, CancellationToken cancellationToken = default)
    {
        await Task.Delay(1);
        if (request == -1)
        {
            throw new Exception("NO -1");
        }
        else
        {
            return "ECHO:" + request.ToString();
        }
    }
}
```

Worker.cs is unmodified from the template.

**Comment by ogiesen** [comment]
This sounds very similar (or at least related) to the issue I reported in #81 . However, that one was auto-closed as stale because nobody responded....

**Comment by jcoliz** [comment]
@ogiesen Nice. It looks like you actually debugged it. I just blamed myself for not figuring it out. Indeed this looks like the same problem. Also I tried it with TCP, and as you predicted that did not work either.

FYI I think you can get the stale tag removed (before an issue is closed) simply by commenting.

**Comment by TaviTruman** [comment]
It seems like you have not registered your MyAsyncHandler in the dependency injection container on the server-side. You need to register the handler in the ConfigureServices method of your server's Program.cs like this:

`services.AddSingleton<IAsyncRequestHandler<int, string>, MyAsyncHandler>();`

Here's the updated Program.cs for the server:


```
using MessagePipe;
using Server;

IHost host = Host.CreateDefaultBuilder(args)
    .ConfigureServices(services =>
    {
        services.AddHostedService<Worker>();
        services.AddMessagePipe();
        services.AddMessagePipeNamedPipeInterprocess("sample-pipe");
        services.AddSingleton<IAsyncRequestHandler<int, string>, MyAsyncHandler>();
    })
    .Build();

host. Run();
```

With these changes, your server should now execute the IAsyncRequestHandler.InvokeAsync method, and the client should receive the response.

**Comment by E-Rooijackers** [comment]
The problem seems to be that the NamedPipeWorker on the server side never gets initialized, thus not listens for connections. NamedPipeWorker will only be initialized by DI when a named pipe based request handler is created, but since no requests can be received, no request handler will get initialized. This is also the case with TCP based request handling.

Workaround for this problem is forcing initialization of NamedPipeWorker/TcpWorker by requesting an instance from the service provider. This way it will start listening for connections and requests can be received.

Also config.HostAsServer should be set to true for server side.

So server side code would then look like:

```
using MessagePipe;
using MessagePipe.Interprocess.Workers;
using Server;

IHost host = Host.CreateDefaultBuilder(args)
    .ConfigureServices(services =>
    {
        services.AddMessagePipe();
        services.AddMessagePipeNamedPipeInterprocess("sample-pipe", config => 
		{	
			config.InstanceLifetime = InstanceLifetime.Singleton;
			config.HostAsServer = true; // Should be true for server side.
		});
        services.AddHostedService<Worker>();
    })
    .Build();
	
var namedPipeWorker = app.Services.GetRequiredService<NamedPipeWorker>();
//tcpWorker = app.Services.GetRequiredService<TcpWorker>(); // Or TcpWorker for TCP based communication

host. Run();

namedPipeWorker.Dispose();
//tcpWorker.Dispose();
```

**Comment by jcoliz** [comment]
@E-Rooijackers @TaviTruman Thanks! I needed to do both of these suggestions, and now it is in fact working.

The only thing I'd add is that getting the `namedPipeWorker` service results in a `ObjectDisposedException` on exit. So I needed to enclose the `GetRequiredService()` *and* `host.Run()` together in a try/catch block.

Anyway, it does seem like this is an actual bug, even though it has a workaround.
